### PR TITLE
SI-6455 no longer rewrite .withFilter to .filter

### PR DESCRIPTION
**Pull request by adriaanm**
This has been deprecated for two major releases now,
and during that time caused plenty of harm (see also SI-7239).

Time to retire, rewrite.

What say ye, @gkossakowski, @Ichoran & @retronym?

### CODE REVIEW

**Comment by retronym** [comment]
The std lib doesn't even fully comply iirc. Was it Either? Maybe scalacheck too?

**Comment by adriaanm** [comment]
Urgh. I admit this is a bit of an eager submission. Deprecation fatigue has set in.

**Comment by adriaanm** [comment]
I'll back off for RC1 if we're indeed this badly broken. That'll teach us ignoring -deprecation.

**Comment by adriaanm** [comment]
Let's see if I can get https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/ to inspire confidence on this one.

**Comment by adriaanm** [comment]
The only class I found that defines filter but not withFilter is Try, but @jorgeortiz85 is [on the case](https://github.com/twitter/util/pull/63/files)!

**Comment by adriaanm** [comment]
I decided to follow the pattern implemented in Option. I manually inspected Scalacheck (and it's compiled by the IDE validation) -- it seems to support withFilter where filter is.

**Comment by adriaanm** [comment]
https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/1/console is building the head of this PR, according to the log:
`"uri" : "git://github.com/scala/scala.git#bc2a1c1d4862bac5c47345b95d8ce486042fe083"`

**Comment by adriaanm** [comment]
Giving the community build another shot at https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/2/console

**Comment by adriaanm** [comment]
Dbuild failed again. I'll defer to 2.12. Should've done this sooner in the cycle.

**Comment by adriaanm** [comment]
I weakened it to make -Xfuture enforce the deprecation.

File: src/library/scala/util/Try.scala [hunk 1]
```diff
@@ -111,6 +111,35 @@ sealed abstract class Try[+T] {
...
+   * As Try is a one-element collection, this may be a bit overkill,
+   * but it's consistent with withFilter on Option and the other collections.
+   *
+   *  @param p   the predicate used to test elements.
+   *  @return    an object of class `WithFilter`, which supports
+   *             `map`, `flatMap`, `foreach`, and `withFilter` operations.
+   *             All these operations apply to those elements of this Try
+   *             which satisfy the predicate `p`.
+   */
+  @inline final def withFilter(p: T => Boolean): WithFilter = new WithFilter(p)
```
Comment (gkossakowski) [inline src/library/scala/util/Try.scala:130]: What does `@inline` achieve here? We won't get rid of the passed closure but just inline the constructor call. I don't think it's worth it (JVM handles this much better than we do).

Comment (adriaanm) [inline src/library/scala/util/Try.scala:130]: Could be, this was copy pasted from Option. Looks like I forgot to note that in the comment.

**Comment by adriaanm** [comment]
test added, comment for Try's withFilter test clarified

**Comment by gkossakowski** [comment]
LGTM

The `@inline` is still there but it won't harm us too much.

**Pull request merged**